src/engine/GroupBy.cpp

```diff
@@ -1010,31 +1006,34 @@ void GroupBy::computeGroupByForHashMapOptimization(
 
   size_t blockSize = 65536;
 
+  // Initialize vectors
+  std::vector<std::vector<AverageAggregationData>> aggregationData;
+  aggregationData.resize(numAggregates);
+
   for (size_t i = 0; i < subresult.size(); i += blockSize) {
     evaluationContext._beginIndex = i;
     evaluationContext._endIndex = std::min(i + blockSize, subresult.size());
 
     auto currentBlockSize =
         evaluationContext._endIndex - evaluationContext._beginIndex;
 
-    // Perform HashMap lookup for all groups in current block
-    std::vector<std::array<AverageAggregationData, numAggregates>*> hashEntries;
-
     auto getId = [&subresult, &evaluationContext, &columnIndex](size_t j) {
       return subresult(evaluationContext._beginIndex + j, columnIndex);
     };
 
-    // Create elements in map
+    // Perform HashMap lookup once for all groups in current block
+    std::vector<size_t> hashEntries;
     for (size_t j = 0; j < currentBlockSize; ++j) {
       auto id = getId(j);
-      map.try_emplace(id);
-    }
+      auto entry = map.try_emplace(id);
 
-    // Get pointers to values
-    // (valid for this iteration as no new elements added)
-    for (size_t j = 0; j < currentBlockSize; ++j) {
-      auto id = getId(j);
-      hashEntries.emplace_back(&(map.at(id)));
+      // New value has been inserted
+      if (entry.second) {
+        entry.first->second = aggregationData.at(0).size();
+        for (auto& aggregation : aggregationData) aggregation.emplace_back();
+      }
```

Comment:
In a data-oriented manner you first go over the hash map and only remember the new indices and the number of entries to insert.
Then AFTER that you go over all the aggregates and do all the `emplace`s at once (via a simple call to `resize` for example).